### ISSUE TITLE
Sharpen only edges needs review and speedup

### ISSUE THREAD

**Issue by heckflosse** [Beep6581/RawTherapee#3656]
As mentioned on pixls.us 'Sharpen only edges` is quite slow. Also the code is [ugly and full of macros](https://github.com/Beep6581/RawTherapee/blob/master/rtengine/bilateral2.h)
There is also at least one bug in the code.

This Issue is to improve this :)

@Floessie and I will work on this without pressure to have another fun optimization session :)
Everybody else is welcome to join!

Ingo

**Comment by heckflosse** [comment]
Here's a measure of current code behaviour. Compared `default sharpening` to `default sharpening + only edges` exporting a full size D800 file (only sharpening part measured)
```
default sharpening                           153 ms
default sharpening + only edges             2856 ms
```

**Comment by Floessie** [comment]
Having had a quick look at the (indeed ugly) code, here is my preliminary plan:

1. Replace macros with proper C++11 code
2. Factor out common code
3. See, if `double` could better be `float` in some places
4. See, if we can vectorize parts

You wrote, you already had some ideas. So maybe I'll branch off from `development` (if that's ready by then, @Beep6581?) or `master` on the weekend and start with 1. and you join me on 2. after that. Okay?

Looking forward to it,
Flössie

**Comment by heckflosse** [comment]
@Floessie Sounds good. :+1: 

btw: the experimental function at the end of bilateral2.h is unused.


Meanwhile here my first ideas:

1. There is a lookup table `ec` of size 512 kb which is created per omp thread. It would be more cache friendly to create this lookup table only once.

2. In the BL_OPER macros the axy parameters can be reduced because always axy == ayx (a12 == a21 and so on).

3. The number of multiplications in the BL_OPERx macros can be reduced by
x*x - ((x+1)/2 * ((x+1)/2+1))/2. That means number of multiplications in for example BL_OPER11 can be reduced by 121 - 21 = 100

4. In BL_BEGIN macro the `scale` and so the macro parameter `a` too can be eliminated because it gets divided out anyway in further processing

**Comment by heckflosse** [comment]
@Floessie 
I had a deeper look at the code.
In principle it divides one gaussian blur by another gaussian blur.
That could lead to a completely new (and much faster) implementation which would make steps 1 to 4 of your plan a waste of time.
I will try that now for the simple 3x3 case...

**Comment by Floessie** [comment]
Ingo, you're the wizard, I'm just a medico. :wink:

**Comment by heckflosse** [comment]
@Floessie Unfortunately I'm not the wizard. I missed some things. :(
You can follow your original plan :+1:

**Comment by heckflosse** [comment]
Before I forget it I would like to mention another possible optimization:

Currently the lookup table needs 512 kb
I inspected the content of the lookup table:
It's symmetrical, the middle value is always 1.0 and it converges to 0.0 at both ends.

That leads to the following possible optimization:

1. Set the lookup table to clip at lower and upper end

2. Calculate the range of the lookup table so it contains no value less then a special epsilon > 0.f. Value of epsilon has to be estimated though).

3. Use this new range to make a lookup table which needs less memory and leads to more cache hits.

Just an idea, but I already tested it and it improves speed for small `Edge tolerance` values while having the same speed for normal to high `Edge tolerance` values

**Comment by iliasg** [comment]
@heckflosse 
Ingo, if the calculation of the LUT is easy you could replace it with on the fly calculations .. no ?.
Is it a gaussian probability distribution ?

**Comment by heckflosse** [comment]
@iliasg Unfortunately the calculation of the LUT includes an exp()
Edit: Perhaps we can beat it by on the fly SSE calculation but I'm not sure

**Comment by Floessie** [comment]
@heckflosse Maybe with some inspiration from [here](http://gruntthepeon.free.fr/ssemath/sse_mathfun.h).

**Comment by heckflosse** [comment]
@Floessie I will test with [this one](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/sleefsseavx.c#L1308) first ;-)

**Comment by heckflosse** [comment]
@Floessie A correct version of [this](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/LUT.h#L292) maybe also handy

**Comment by iliasg** [comment]
@heckflosse Ingo, is the current implementation of bilateral filter optimized for speed for the given task ?. 

As I understand, full accurate (exact) Bilateral filters are generally very slow, but fast bilaterals have been around optimized for speed by decreasing precision. Some are effective (speed/quality) on small radius only, others are more effective for large radius etc.    
http://people.csail.mit.edu/sparis/publi/2006/tr/Paris_06_Fast_Bilateral_Filter_MIT_TR.pdf
http://people.csail.mit.edu/sparis/bf/

**Comment by heckflosse** [comment]
@iliasg The source code has this comment:
```
// This seems ugly, but way faster than any other solutions I tried
```
So it looks like someone optimized it long time ago. 

It also has some relics from old times where rt processing was integer (for example the gauss approximation using large integral coefficients). I am not sure whether we should change this or not.

In my tests I could already reduce the processing time for the case mentioned in second post from 2856 ms to ~1800 ms without using SSE. With some use of  SSE code I expect an about 2x speedup from the current ~1800 ms to ~900 ms for a D800 file.

So I wouldn't say it's very slow, but it is in fact the brute force implementation described on page 12 of the paper you mentioned.

I suggest to follow what @Floessie suggested to get an identical output (except for the two bugs in current code) at reduced processing time in first step.
In a second step we can try to use a faster algorithm.

**Comment by iliasg** [comment]
@heckflosse 
OK :)
Just in case you will use sleef.c .. it's updated to 2.121
The log says for 2.110, "Following functions returned incorrect values when the argument is very large or small : exp, pow, asinh, acosh"
https://github.com/shibatch/sleef

**Comment by heckflosse** [comment]
@iliasg 
Ilias, for the incorrect values of exp in sleef there's a workaround in rt version of sleef since long time :)

sleef pow, asinh and acosh currently are not used by rt

I prefer to not change a running system a the moment ;-)

**Comment by Floessie** [comment]
@heckflosse I've just pushed a [new branch](https://github.com/Beep6581/RawTherapee/tree/bilateral_speedup) as the base for further experimentation. The only properly converted macro is [`BL_OPER3`](https://github.com/Beep6581/RawTherapee/blob/bilateral_speedup/rtengine/bilateral.cc#L234) for now. Maybe we start optimizing that and apply the rules we find to the other variants afterwards.

**Comment by heckflosse** [comment]
@Floessie Shall I just push some changes?

**Comment by heckflosse** [comment]
@Floessie I just pushed some small changes which also should help to understand the code.
Only optimization I made was one LUT instead of one LUT per thread.

**Comment by heckflosse** [comment]
@Floessie To avoid that I forget it:
Currently the bilateral function has 3 arrays (source, destination and buffer)
But buffer is only needed when `source == destination`.
I suggest to allocate the buffer inside the bilateral function for the case `source == destination` and to set
`buffer = destination;` for the case `source != destination`
That would also mean in case `source == destination` we don't have to copy the borders from source to destination and in case `source != destination` we only have to copy the borders from source to destination in last step.

**Comment by Floessie** [comment]
@heckflosse But the buffer is needed by `deconvsharpening()` and `gaussianBlur()`, too.

**Comment by heckflosse** [comment]
@Floessie Yes, b2 is used for gaussianBlur and deconv...
But we don't need it for bilateral currently. We can directly write into b3
```
Bilateral(lab->L, (float**)b3, b3, W, H, sharpenParam.edges_radius / scale, sharpenParam.edges_tolerance, ec).compute();
```
;-)

**Comment by heckflosse** [comment]
@Floessie Do you have some changes on hold or can I push some changes today?

**Comment by Floessie** [comment]
@heckflosse No, you can push. I was busy with my color negatives.

**Comment by heckflosse** [comment]
@Floessie I pushed two commits. No speedups, but fixed two old bugs and cleaned the new code a bit

**Comment by Floessie** [comment]
@heckflosse Nice changes. I wasn't really happy with having `ec` outside the class. That's much better. I'll give it some finishing now.

**Comment by heckflosse** [comment]
@Floessie :+1:

**Comment by Floessie** [comment]
Something's wrong with `LUT`: It doesn't have a move-ctor but elements can be changed although it's `const`. :unamused: 

I'll convert the other `BL_OPER` macros during this week. Feel free to push in the meantime.

When calling `Bilateral::Bilateral()` there is an needless explicit `float**` cast for `b3`. Missed it in my commit.

**Comment by heckflosse** [comment]
@Floessie The size of ` ec ` will be variable in future.
Edit: variable in dependence of ` sensitivity `

**Comment by Floessie** [comment]
@heckflosse Surprisingly, `bilateral07()` and `bilateral09()` share the same parameters. Bug?

**Comment by heckflosse** [comment]
@Floessie Definitely an old (copy/paste) bug.

**Comment by heckflosse** [comment]
@Floessie I think we should get rid off these old integer based gaussian coefficents. I will calculate new floating point coefficients or (even better) we should calculate them dynamically based on sigma!

**Comment by iliasg** [comment]
@heckflosse .. +2 ;) 
.. better accuracy, less memory .. maybe same or better speed ?.

**Comment by heckflosse** [comment]
@iliasg 
better accuracy? Yes
less memory? Peak memory usage will not be reduced I think
same or better speed? We will see

@Floessie May I push some changes today?

**Comment by Floessie** [comment]
@heckflosse Yes, please. :)

**Comment by heckflosse** [comment]
@Floessie I pushed 2 commits

**Comment by heckflosse** [comment]
@Floessie 

Some explanations about the changes:

1. by nature of ec the middle of the ec lut always has value 1. Also for the centre point of the grid always ec[middle point] is used. For that reason I could simplify the calculations with centre point.

2. I wanted to reflect the names of the gauss kernel values to the distance from centre value for better readability of code (c00 means 0 distance to centre in both directions, c01 means distance 1 in one direction and distance 0 in other direction and so on)

3. I reduced the number of multiplications

4. not using suly in elem gave the largest speedup

Ingo

**Comment by erickpaquin** [comment]
@heckflosse , should this PR be closed now?

**Comment by Thanatomanic** [comment]
Not Ingo, but I'm sure that this issue is not really relevant anymore. Maybe in the future, there will be more optimizations.
